libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -471,6 +490,180 @@ uint64_t VoteManager::roundDeterminedFromVotes(size_t two_t_plus_one) {
+    return false;
+  }
+
+  if (!voter_dpos_votes_count) {
+    LOG(log_er_) << "Invalid reward vote: voter was not eligible to vote " << *vote;
+    return false;
+  }
+
+  const uint64_t reward_period_dpos_total_votes_count = final_chain_->dpos_eligible_total_vote_count(dpos_period);
+  const size_t reward_period_pbft_sortition_threshold =
+      db_->getPbftSortitionThreshold(dpos_period + 1);  // reward period
+  if (!reward_period_pbft_sortition_threshold) {
+    LOG(log_er_) << "Cannot get PBFT sortition threshold for period " << dpos_period;
+    return false;
+  }
+  try {
+    vote->validate(voter_dpos_votes_count, reward_period_dpos_total_votes_count,
+                   reward_period_pbft_sortition_threshold);
+  } catch (const std::logic_error& e) {
+    LOG(log_er_) << e.what();
+    LOG(log_er_) << "DPOS period " << dpos_period << ", voter " << voter_account_addr << " votes count "
+                 << voter_dpos_votes_count << ", total votes count " << reward_period_dpos_total_votes_count
+                 << ", pbft sortition threshold " << reward_period_pbft_sortition_threshold;
+
+    return false;
+  }
+
+  return true;
+}
+
+void VoteManager::updateRewardVotes(uint64_t reward_period) {
+  auto period_data = db_->getPeriodDataRaw(reward_period);
+  if (period_data.size() == 0) {
+    // First period no reward votes
+    return;
+  }
+
+  SyncBlock reward_period_sync_block(period_data);
+
+  std::unordered_set<vote_hash_t> reward_period_cert_votes_hashes;
+  for (const auto& v : reward_period_sync_block.cert_votes) {
+    reward_period_cert_votes_hashes.insert(v->getHash());
+  }
+
+  bool update = false;
+  {
+    const auto reward_period_pbft_block_hash = reward_period_sync_block.pbft_blk->getBlockHash();
+    std::unique_lock lock(reward_votes_mutex_);
+    if (reward_votes_.first != reward_period_pbft_block_hash) {
+      LOG(log_dg_) << "Reward period pbft block<" << reward_period_pbft_block_hash
+                   << "> does not match reward votes voted value " << reward_votes_.first;
+      return;
+    }
+    for (auto& v : reward_votes_.second) {
+      if (!reward_period_cert_votes_hashes.contains(v.first) && verifyRewardVote(v.second)) {
+        LOG(log_nf_) << "Add new reward vote " << v.first;
+        update = true;
+        reward_period_cert_votes_hashes.insert(v.first);
+        reward_period_sync_block.cert_votes.push_back(std::move(v.second));
+      }
+    }
+    // Clear reward votes table
+    reward_votes_ = {};
+  }
+
+  if (update) {
+    db_->UpdateCertVotesInPeriodData(reward_period_sync_block);
+  }
+}
+
+bool VoteManager::checkRewardVotes(const std::shared_ptr<PbftBlock>& pbft_block) {
+  const auto reward_period = pbft_block->getPeriod() - 1;
+  if (!reward_period) {
+    // First period no reward votes
+    return true;
+  }
+
+  const auto reward_period_cert_votes = db_->getCertVotes(reward_period);
```

Comment:
do not load votes from db it is not necessary and not efficient. You can save hashes of cert votes after new pbft block is saved into the period data and read it here instead of db read

Reply:
Fixed